eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs

```diff
@@ -396,16 +396,26 @@ ledgerTransition = do
               , certState
               , StrictSeq.fromStrict $ txBody ^. certsTxBodyL
               )
-        let wdrlAddrs = Map.keysSet . unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL
-            wdrlCreds = Set.map raCredential wdrlAddrs
-            dUnified = dsUnified $ certDState certStateAfterCERTS
-            delegatedAddrs = DRepUView dUnified
-
-        -- TODO: Finish this implementation once we are in bootstrap phase:
-        -- https://github.com/IntersectMBO/cardano-ledger/issues/4092
-        when False $ do
-          all (`UMap.member` delegatedAddrs) wdrlCreds
-            ?! ConwayWdrlNotDelegatedToDRep (wdrlCreds Set.\\ Map.keysSet (dRepMap dUnified))
+
+        let checkWdrlsKeyHashesDelegated = do
+              let
+                isKeyHash = \case
+                  KeyHashObj _ -> True
+                  _ -> False
+                wdrlsKeyHashes =
+                  Set.filter isKeyHash $
+                    Set.map raCredential $
+                      Map.keysSet $
+                        unWithdrawals $
+                          tx ^. bodyTxL . withdrawalsTxBodyL
+                dUnified = dsUnified $ certDState certStateAfterCERTS
+                notDelegated = wdrlsKeyHashes Set.\\ Map.keysSet (dRepMap dUnified)
+              Set.null notDelegated ?! ConwayWdrlNotDelegatedToDRep notDelegated
```

Comment:
Is there a difference between: 
 * checking that each credential is  `Umap.member` of `DRepUView` and
 *  just looking at the difference between  the credentials and `dRepMap`  ? 
 
 After staring at the code I couldn't find any, but on the other hand there might be a good reason for which this was implemented like that in the first place?  @aniketd @Soupstraw  what do you think?

Reply:
They should be the same, I think the reason why we look at the difference here is because we want to include the notDelegated set in the failure.

Reply:
They will have the same outcome, but checking the membership will be a whole lot faster. 

The way `dRepMap` works is that it takes a huge map with millions of entries and then filters out all of the ones that have delegated to a DRep. That is an `O(n*log n)` operation.

Then you are doing another restriction with `k` withdrawals that are much smaller, and it turns out `O(k*n*log n)` operation. 

On the other hand if you do just a membership check for those `k` withdrawals you get: `O(k*log n)` operation, which will be much faster!

Reply:
Thanks for the generous explanation!